### ISSUE TITLE
No preemption with requested cores

### ISSUE THREAD

**Issue by ferchault** [hashicorp/nomad#12712]
### Nomad version
Nomad v1.2.6 (a6c6b475db5073e33885377b4a5c733e1161020c)

### Issue
We try to use [preemption](https://learn.hashicorp.com/tutorials/nomad/preemption) based on number of cores using this minimal job file where the goal is to reserve 10 cores for one task and preempt any jobs that might interfere. Asking for one group of 10 cores yields a placement error, asking for 10 groups of one core does preempt the running jobs.

### Reproduction steps
```
job "redis" {
  datacenters = ["dc1"]
  type        = "service"
  priority    = 80

  constraint {
    attribute = "${node.unique.id}"
    value = "0dca1233-20b6-455d-815f-af615090d533"
  }

  group "cache1" {
    count = 1

    task "redis" {
      driver = "raw_exec"
				user = "leruli"

      config {
        command = "/bin/bash"
        args = ["sleep", "10000"]
      }
      
      resources {
        memory = 100
        cores = 10
      }
    }
  }
}
```
This yields a failed placement for any number of cores > 1 even though the node is fully allocated and no core is available:
![image](https://user-images.githubusercontent.com/2736606/164126221-2907e52d-f546-4352-9e15-18a2c4a95618.png)

The strange thing is asking for 10 instances of a group of one core works as expected with a set of jobs marked for preemption:
```
job "redis" {
  datacenters = ["dc1"]
  type        = "service"
  priority    = 80

  constraint {
    attribute = "${node.unique.id}"
    value = "0dca1233-20b6-455d-815f-af615090d533"
  }

  group "cache1" {
    count = 10

    task "redis" {
      driver = "raw_exec"
      user = "leruli"

      config {
        command = "/bin/bash"
        args = ["sleep", "10000"]
      }
      
      resources {
        memory = 100
        cores = 1
      }
    }
  }
}
```
![image](https://user-images.githubusercontent.com/2736606/164127047-e8f09a65-c558-49ab-9714-d2b4be71154a.png)

#### Expected Result
Asking for 10 cores on a specific node should preempt other jobs regardless whether its one job of 10 cores or 10 jobs of 1 core.

#### Actual Result
Only asking for 10 tasks of one core each preempts.

**Comment by tgross** [comment]
Hi @ferchault! That definitely seems like a bug. I took a quick glance at the [preemption](https://github.com/hashicorp/nomad/blob/main/scheduler/preemption.go) code and it looks like the cores are included in the task resources we're checking, but maybe we're not calculating the score correctly for them. We'll need to write some new test code to verify where the issue is.

Another thing that caught my eye when reviewing the docs was  [priority delta](https://www.nomadproject.io/docs/internals/scheduling/preemption#details):

> In order to prevent a cascade of preemptions due to jobs close in priority being preempted, only allocations from jobs with a priority delta of more than 10 from the job needing placement are eligible for preemption.

But I don't think that's it either. The job with 10 allocs with 1 core each uses 10x as much memory, so I suspect the reason you're seeing the difference in the behavior is because we're preempting on memory correctly, but not cores. If the problem was the priority delta, we wouldn't be seeing preemption at all.

All that being said, is your real workload using `raw_exec`? The scheduler should be able to make appropriate placements with that (including preemption), but there's no resource isolation with `raw_exec`, so reserving CPUs doesn't actually do anything on the client.

**Comment by ferchault** [comment]
> But I don't think that's it either. The job with 10 allocs with 1 core each uses 10x as much memory, so I suspect the reason you're seeing the difference in the behavior is because we're preempting on memory correctly, but not cores. If the problem was the priority delta, we wouldn't be seeing preemption at all.

Agreed. I checked, the other jobs run with priority 50. These jobs are CPU-bound, so the node has enough memory such that preemption should not be required for memory at all.

> All that being said, is your real workload using `raw_exec`? The scheduler should be able to make appropriate placements with that (including preemption), but there's no resource isolation with `raw_exec`, so reserving CPUs doesn't actually do anything on the client.

Yes, we enforce resource isolation within each job ourselves, relying on nomad for the resource scheduling only.
